### DISCUSSION TITLE
Correct way of integrate flux into an existing cluster

### DISCUSSION THREAD

**Discussion by Thaval** [fluxcd/flux2#2718] [Q&A]
Hi, 

I'm struggling a little bit getting flux into our cluster flawlessly. 
So, currently I annotate all resources created by myself with the appropriate labels and annotations as suggested in the docs, to tell flux "manage these resources". 
Then I create my GitRepository using the Flux CLI and the HelmRelease using the Flux CLI. Unfortunately, I cannot configure things like "rollback", "install" or "upgrade" through the CLI.
But the HelmRelease cannot be created. There are different errors coming out, all result in "upgrade retries exhausted.".
Most of the times it complains about that an "env" variable is invalid as it is empty (I'm binding a lot of value from Secrets or ConfigMaps).
The crazy thing is: When I delete the HelmRelease (all resources are deleted too) and re-create the HelmRelease, the deployment then works. 
However, as I want to integrate Flux also for our productive systems, this is not a feasible solution to me. 

Is there something I missed regarding to Flux? Because it seems that my Helm setup is correct, otherwise I wouldn't be able to install it when the workspace is cleanup after deleting the HelmRelease.

**Comment by jprecuch** [comment]
@Thaval it always depends what the error(s) is when the FluxCD helm install happens ("There are different errors coming out"). If you fix those errors that are thrown by FluxCD doing helm install then FluxCD should work fine and without issues.

**Reply by Thaval** [reply]
It says 
"[spec.template.spec.containers[0].env[5].valueFrom: Invalid value: "": may not have more than one field specified at a time, spec.template.spec.containers[0].env[9].valueFrom: Invalid value: "": may not have more than one field specified at a time]"
But only if the resources are already there. When I deploy it from scratch (so the workloads are not existing), then the flux deployment works.

Oh, and now I have found another issue: "spec.selector: immutable field"

**Reply by Thaval** [reply]
I just found out that such a change will cause these "Invalid value" errors
```
# current running deployment
- name: EXPRESS_LOG_LEVEL
  value: debug

# You are trying to update to
- name: EXPRESS_LOG_LEVEL
  valueFrom:
    configMapKeyRef:
      name: xxx-configmap
      key: EXPRESS_LOG_LEVEL
```

I guess it's a bug then?

**Reply by kingdonb** [reply]
I wonder if the values are the same or different? If Helm values are the same then Flux's helm controller should behave the same as Helm. Do you know about `helm get values [releasename]` to retrieve the `values` from an existing release?

You can use this feature to compare existing and new value sets, to ensure that the transition is as seamless as possible.

If you are still affected by this issue and can document the steps to reproduce it, please file it as an issue under helm-controller. If there is a bug here then I haven't been specifically made aware of it before, but it is possible there is a bug.
